src/library/scala/collection/immutable/HashMap.scala

```diff
@@ -322,27 +333,31 @@ object HashMap extends ImmutableMapFactory[HashMap] with BitOperations.Int {
     override protected def contains0(key: A, hash: Int, level: Int): Boolean =
       hash == this.hash && key == this.key
     private[collection] override def updated0[B1 >: B](key: A, hash: Int, level: Int, value: B1, kv: (A, B1), merger: Merger[A, B1]): HashMap[A, B1] =
-      if (hash == this.hash && key == this.key) {
-        if (merger eq null) {
-          if (this.value.asInstanceOf[AnyRef] eq value.asInstanceOf[AnyRef]) this
-          else new HashMap1(key, hash, value, kv)
-        } else {
-          val current = this.ensurePair
-          val nkv = merger(current, if (kv != null) kv else (key, value))
-          if ((current eq nkv) || (
-            (current._1.asInstanceOf[AnyRef] eq nkv._1.asInstanceOf[AnyRef]) &&
-              (current._2.asInstanceOf[AnyRef] eq nkv._2.asInstanceOf[AnyRef]))) this
-          else new HashMap1(nkv._1, hash, nkv._2, nkv)
-        }
-      } else {
-        if (hash != this.hash) {
-          // they have different hashes, but may collide at this level - find a level at which they don't
-          val that = new HashMap1[A, B1](key, hash, value, kv)
-          makeHashTrieMap[A, B1](this.hash, this, hash, that, level, 2)
-        } else {
-          // 32-bit hash collision (rare, but not impossible)
+      if (hash == this.hash) {
+        if (key == this.key) {
+          if (merger eq null) {
+            if (this.value.asInstanceOf[AnyRef] eq value.asInstanceOf[AnyRef]) this
+            else new HashMap1(key, hash, value, kv)
+          } else if (
+            (key.asInstanceOf[AnyRef] eq this.key.asInstanceOf[AnyRef]) &&
+              (value.asInstanceOf[AnyRef] eq this.value.asInstanceOf[AnyRef]) &&
+              merger.retainIdentical) {
+            this
+          } else {
+            val current = this.ensurePair
+            val nkv = merger(current, if (kv != null) kv else (key, value))
+            if ((current eq nkv) || (
+              (key.asInstanceOf[AnyRef] eq nkv._1.asInstanceOf[AnyRef]) &&
+                (value.asInstanceOf[AnyRef] eq nkv._2.asInstanceOf[AnyRef]))) this
```

Comment:
Doesn't this need to be guarded by `merger.retainIdentical` but can be removed because it was already covered by the previous branch anyway?

Reply:
I don't see why. 
If the merge generates a tuple which is identical in all respects then we don't need to create a new `HashMap1`, when this is a clone and identical in all respects to `this`
What user visible difference is there? 
Non-functionally there is a change because not having this would cause the creation of a new `HashMap1` and all of the parent structures, and the loss of structural sharing

Reply:
I think I see the confusion - there was a bug here
key and values are updated to be this.key and this.value - no referring to the parameters

Reply:
Ah, that explains it. I was looking for the context of the code again after your previous comment but github didn't have it anymore so I looked at the latest version and everything there looked fine.